pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+// signing CertificateSigningRequests that reference a cert-manager Venafi
+// Issuer or ClusterIssuer
+type Venafi struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+	certClient    certificatesclient.CertificateSigningRequestInterface
+	recorder      record.EventRecorder
+
+	clientBuilder venaficlient.VenafiClientBuilder
+}
+
+func init() {
+	controllerpkg.Register(CSRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		return controllerpkg.NewBuilder(ctx, CSRControllerName).
+			For(certificatesigningrequests.New(apiutil.IssuerVenafi, NewVenafi(ctx))).
+			Complete()
+	})
+}
+
+func NewVenafi(ctx *controllerpkg.Context) *Venafi {
+	return &Venafi{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		certClient:    ctx.Client.CertificatesV1().CertificateSigningRequests(),
+		recorder:      ctx.Recorder,
+		clientBuilder: venaficlient.New,
+	}
+}
+
+// Sign attempts to sign the given CertificateSigningRequest based on the
+// provided Venafi Issuer or ClusterIssuer. This function will update the resource
+// if signing was successful. Returns an error which, if not nil, should
+// trigger a retry.
+// Since this signer takes some time to sign the request, this controller will
+// set a "pick ID" annotation value that is used to fetch the latest state of
+// the request in subsequent re-syncs.
+func (v *Venafi) Sign(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, issuerObj cmapi.GenericIssuer) error {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	resourceNamespace := v.issuerOptions.ResourceNamespace(issuerObj)
+
+	client, err := v.clientBuilder(resourceNamespace, v.secretsLister, issuerObj)
+	if apierrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+		v.recorder.Event(csr, corev1.EventTypeWarning, "SecretNotFound", message)
+		log.Error(err, message)
+		return nil
+	}
+
+	if err != nil {
+		message := fmt.Sprintf("Failed to initialise venafi client for signing: %s", err)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorVenafiInit", message)
+		log.Error(err, message)
+		return err
+	}
+
+	var customFields []venafiapi.CustomField
+	if annotation, exists := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey]; exists && annotation != "" {
+		err := json.Unmarshal([]byte(annotation), &customFields)
+		if err != nil {
+			message := fmt.Sprintf("Failed to parse %q annotation: %s", experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey, err)
+			v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", message)
+			util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", message)
+			_, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+			return err
```

Comment:
nit: it is hard to "see" whether the initial error is returned or not. At first, I read:

```go
		if err != nil {
			// ...

			return err // Seems like we are retrying by returning this err?
		}
```

then I saw the UpdateStatus in the middle which made me read it like this:


```go
		if err != nil {
		    // ...
		    
			_, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
			if err != nil {
			    return err
            }
            
			return nil // Clearly show that we don't retry the initial err.
		}
```

I'd suggest going with the second approach. It is 3 lines more, but it clearly shows that  the Unmarshall error is never returned.

Reply:
I disagree, but will use a different error variable anyway instead to not need these extra lines.